### PR TITLE
ENH: Significant speedup for ortho and special ortho group

### PR DESCRIPTION
**Pull request by JesseLivezey**
This PR makes 2 changes to `ortho_group.rvs()` and `special_ortho_group.rvs()` that significantly speed up the calculation.

1) The former `Hx` matrix is an identity minus a single outer product. In the next step it is matrix multiplied. It is faster to re-order this process as H.dot(I - xx^t) --> H - H.dot(x)x^T. In this re-ordering, you never have to do a matrix-matrix multiply.
2) The former `mat` matrix was block diagonal with the identity from 0 to n and `Hx` as the second block. On every loop, it was matrix multiplied by by `H`. The identity block means that you can just do the computation on a smaller and smaller subset of the columns of `H` at every iteration.

This changes the scaling of the computation and so it gives large speedups for `dim` > 100.

ortho_group2

special_ortho2

@rgommers, this is a follow-up to #10058. It turns out I missed the most obvious ways to speed these function up in the first pass.

### CODE REVIEW

File: scipy/stats/_multivariate.py [hunk 1]
```diff
@@ -3394,10 +3394,7 @@ def rvs(self, dim, size=1, random_state=None):
             x[0] += D[n]*np.sqrt(norm2)
             x /= np.sqrt((norm2 - x0**2 + x[0]**2) / 2.)
             # Householder transformation
-            Hx = (np.eye(dim-n) - np.outer(x, x))
-            mat = np.eye(dim)
-            mat[n:, n:] = Hx
-            H = np.dot(H, mat)
+            H[:, n:] = H[:, n:] - np.outer(np.dot(H[:, n:], x), x)
```
Comment (ilayn) [inline scipy/stats/_multivariate.py]: Wouldn't you use `-=` here instead?

Comment (JesseLivezey) [reply scipy/stats/_multivariate.py]: Yep, I'll wait for the tests to finish and update.

File: scipy/stats/_multivariate.py [hunk 2]
```diff
@@ -3394,10 +3394,7 @@ def rvs(self, dim, size=1, random_state=None):
             x[0] += D[n]*np.sqrt(norm2)
             x /= np.sqrt((norm2 - x0**2 + x[0]**2) / 2.)
             # Householder transformation
-            Hx = (np.eye(dim-n) - np.outer(x, x))
-            mat = np.eye(dim)
-            mat[n:, n:] = Hx
-            H = np.dot(H, mat)
+            H[:, n:] -= np.outer(np.dot(H[:, n:], x), x)
```
Comment (ev-br) [inline scipy/stats/_multivariate.py]: Does it need `np.outer`? It's scalar-vector product, is it not?

Comment (JesseLivezey) [reply scipy/stats/_multivariate.py]: Unless I'm missing another simplification, I think this does need outer. The result of `np.dot(H[:, n:], x)` is a vector and you want the outer product of that vector and `x`.

**Comment by JesseLivezey** [comment]
Some added detail for potential future readers:

Previously, there was a `(dim x dim) x (dim x dim)` matrix multiply in a loop of size `dim`. The complexity of this is O(`dim^4`) assuming square matrix multiplies are O(`dim^3`).

With this change, there is a `(dim x n) x n` matrix-vector product (O(`dim x n`)) followed by a `n x n` vector-vector outer product (O(`n x n`)) where `n` is the loop index inside a loop of size `dim`. Overall, this should be O(`dim^3)`.

These changes just re-order the previous calculations and skip some trivial computations (multiplication by identity, etc.). Up to small numerical differences, the results of this should be identical to the previous version.

I haven't tried this yet, but these changes might also make it possible to vectorize over the `size` argument rather than looping over it (and turn matrix-vector back into matrix-matrix operations).

Unrelated to this PR, I've rewritten this code in pytorch to run on GPUs. I need to generate 1000s of samples with 1000s of dimensions and see pretty good (~25x) speedups for large dims.

https://github.com/JesseLivezey/pytorch_group_sampler

**Review by larsoner — APPROVED** [review]
LGTM +1 for merge. Marked for 1.3.0 because it looks like a reasonable quick thing to merge

**Review by tylerjereddy — COMMENTED** [review]
@JesseLivezey I'm trying to run the `asv` benchmarks on the performance improvements here. I noticed that there's a small problem, which isn't your fault, but if you could push this small adjustment in that would be helpful, as the related benchmark has a missing argument in its `setup()`


```diff
diff --git a/benchmarks/benchmarks/stats.py b/benchmarks/benchmarks/stats.py
index 780550001..d6a37b2ed 100644
--- a/benchmarks/benchmarks/stats.py
+++ b/benchmarks/benchmarks/stats.py
@@ -141,7 +141,7 @@ class GroupSampling(Benchmark):
     param_names = ['dim']
     params = [[3, 10, 50, 200]]

-    def setup(self):
+    def setup(self, dim):
         np.random.seed(12345678)

     def time_unitary_group(self, dim):
```

This can also be seen by the almost completely blank performance results plots for `ortho_group` at our benchmark page: https://pv.github.io/scipy-bench/  & would help preserve the performance improvements here moving forward.

**Comment by JesseLivezey** [comment]
@tylerjereddy Sadly, I am to blame since I added that (broken) benchmark in a recent PR. :) I pushed the fix.

Looking at the scipy `asv`, I noticed that another unrelated benchmark I added a while ago doesn't have any data
https://pv.github.io/scipy-bench/#spatial.Radius.time_query_ball_point_nosort
which is here
https://github.com/scipy/scipy/blob/8532ca64d5a76e2a014ea1787cd72b79087a5696/benchmarks/benchmarks/spatial.py#L130

Is it missing something obvious? I've never actually used `asv`.

**Review by tylerjereddy — APPROVED** [review]
Ok, `asv` benchmarks look solid now:

`asv continuous -e -b "time_ortho_group" master ortho_group_speedup2 >& log.txt`

```
       before           after         ratio
     [8532ca64]       [51b44f30]
     <master>         <ortho_group_speedup2>
-      68.1±0.5μs       59.8±0.2μs     0.88  stats.GroupSampling.time_ortho_group(3)
-       227±0.6μs          194±1μs     0.85  stats.GroupSampling.time_ortho_group(10)
-     1.91±0.03ms      1.28±0.04ms     0.67  stats.GroupSampling.time_ortho_group(50)
-        50.0±6ms       18.6±0.3ms     0.37  stats.GroupSampling.time_ortho_group(200)
```
`asv continuous -e -b "time_special_ortho_group" master ortho_group_speedup2 >& log2.txt`
```
       before           after         ratio
     [8532ca64]       [51b44f30]
     <master>         <ortho_group_speedup2>
-     1.83±0.04ms      1.18±0.01ms     0.64  stats.GroupSampling.time_special_ortho_group(50)
-        54.7±2ms       15.8±0.8ms     0.29  stats.GroupSampling.time_special_ortho_group(200)
```

Thanks for fixing that.

The pypy failure is unrelated, and there's a +1 to merge from a core dev. In it goes, thanks @JesseLivezey

**Comment by tylerjereddy** [comment]
If you could add a small entry summarizing the improvements from this PR at the release note page that will help me out: https://github.com/scipy/scipy/wiki/Release-note-entries-for-SciPy-1.3.0

Glancing at the other benchmark you mention, I don't see an immediately obvious problem. You could always try running it with a command similar to what I've used above.

**Comment by JesseLivezey** [comment]
@tylerjereddy I added a comment. Thanks!

**Pull request merged**